### DISCUSSION TITLE
Question on how prediction / tokenization works

### DISCUSSION THREAD

**Discussion by vade** [openai/whisper#832] [Q&A] [answered]
Hello

Im working on a port of Whisper to CoreML / vDSP / Accelerate, and am at a place where im getting sentence output but its non sensical. Im trying to understand what im doing wrong. Any feedback / guidance would be appreciated! If you want to follow along, the code is here:

https://github.com/tanmayb123/OpenAI-Whisper-CoreML/pull/2

What I have so far:
* Accelerate based Log Mel spectrogram creation that numerically is close, but not identical to the PyTorch data loading / audio pre processing, and using the same pre-calculated/cached filter bank
* Converted CoreML models. The decoder has a flexible input shape for tokens (ie it can take 1 x n token inputs)
* An attempt (ha) at re implementing the Whisper Tokenizer, using the Hugging face Swift GPT 2 tokenizer as a base class.

What I am doing with the components:

* Accrue a frame's worth of samples (480,000), ie 16Khz x 30seconds
* Generate the Log Mel
* Encode the Log Mel into a 1 x 80 x 3000 tensor (MLMultiArray) producing encoded log mel spectrogram.
* Init a SOT Token sequence, using: [Sot Token, Sot Token + 1 (English?), Transcribe Token]
* Send the SOT Sequence (as a Int 32 ML Multi array) and the encoded log Mel spectrogram to the decoder model.
* This produces a logit tensor for each tokens, shaped 1 x Num Tokens x 51865, repressing the confidence value of each possible token ID for every submitted token.

This is where I lose track of things.

* I am slicing the last tokens logits, ie 1 x Numtokens - 1 to get a 1 x 51865 logit tensor representing the confidence values of the MOST RECENTLY submitted token
* I run softmax on those logits, and mask / remove any special tokens and their confidences. This leaves me with a specific token index and its confidence.
* I then append this token to an array of previously extracted tokens (for the first run, my SOT sequence from above, plus the latest)
* I repeat this process until I hit the EOT token 
* I can then decode via the tokenizer to get my sentence.

Questions:

a) is this more or less correct?
b) in the [PyTorch Inference class](https://github.com/openai/whisper/blob/main/whisper/decoding.py#L141), it seems like the `logits` method is actually NOT submitting a list of already found tokens, but just the last token? "# only need to use the last token except in the first forward pass" - If I do that in my code base, I only get the same token repeated ad infinitum.

Any guidance is MUCH appreciated. Getting this working would allow for very lightweight and power efficient iOS and macOS transcription due to the underlying use of native hardware and SIMD extensions. 

Much obliged!

**Comment by vade** [comment, marked as answer]
Hi. 

Answering my own questions here. The above is correct. My code works if I use a MEL Spectrogram imported from PyTorch. The CoreML code, encoding and decoding all work as expected, which is awesome. Now I just need to figure out why my MEL code is producing incorrect outputs compared to Pytorch's STFT implementation

Below is a pure MacOS, Accelerate and CoreML powered whisper small model running with only 1 core and the ANE 

Still a long way to go, but its getting there:

![image](https://user-images.githubusercontent.com/65011/211876759-d753b58a-547e-40d3-986a-17ab24572a3c.png)

**Comment by tom-huntington** [comment]
> in the [PyTorch Inference class](https://github.com/openai/whisper/blob/main/whisper/decoding.py#L141), it seems like the logits method is actually NOT submitting a list of already found tokens, but just the last token? "# only need to use the last token except in the first forward pass" - If I do that in my code base, I only get the same token repeated ad infinitum.

There is caching involved here. The keys and values of tokens don't change as more tokens are appended. So only the keys and values are calculated for the last token. However, keys and values for all the tokens are needed for the forward pass.

https://github.com/openai/whisper/blob/f82bc59f5ea234d4b97fb2860842ed38519f7e65/whisper/decoding.py#L134

**Reply by vade** [reply]
Thanks Tom -  I dont think CoreML can use those Keys and values directly (or at least, its not obvious to me how). What im doing is caching the tokens that are calculated from the last pass, and submitting that to the forward / inference pass, which as I accrue more tokens makes prediction seem to take longer.

Forgive me for a maybe naive q - but does the `self.kv_cache` in PyTorch keep the forward pass runtime the same the further into token prediction? (ie, as the cache grows?) or is there a similar slow down?

**Reply by tom-huntington** [reply]
I think a slow down should be expected. The size of the cross attention matrix changes, unlike with the language models which are padded with blank tokens.

Although I'm more confused than an expert

**Reply by tom-huntington** [reply]
I'm just using this guys work https://github.com/zhuzilin/whisper-openvino

**Reply by vade** [reply]
Haha, you and me both!
